### PR TITLE
refactor/bugfix: move CanSendHail to ship

### PR DESCRIPTION
**Pull request by Hurleveur**
**Bugfix:** prevents secretive provoked on scan ships from hailing the player if they are mute or if they do not have the same government

## Fix Details
The CanSendHail method was located in the nameless namespace of Engine (dont ask me why) I moved it to ship so we can check before hailing the player to warn them we'll shoot them (or not)

## Testing Done
if the tests run this is good; the code for CanSendHail is already used all the time, and there are checks within it.
and no referencing nullptr here (I added a check each time)

## Save File
n/a

### CODE REVIEW

**Review by RisingLeaf — APPROVED** [review]
looks good

File: source/Ship.cpp [hunk 1]
```diff
@@ -2671,7 +2703,8 @@ int Ship::Scan()
 			Messages::Add("Attempting to scan the selected " + target->Noun() + "."
 				, Messages::Importance::Low);
 
-		if(target->GetPersonality().IsSecretive() && target->GetGovernment()->IsProvokedOnScan())
+		if(target->GetPersonality().IsSecretive() && target->GetGovernment()->IsProvokedOnScan()
+			&& target->CanSendHail(player))
```
Comment (warp-core) [inline source/Ship.cpp:2707]: Shouldn't this be:
```suggestion
		if((target->GetPersonality().IsSecretive() || target->GetGovernment()->IsProvokedOnScan())
			&& target->CanSendHail(player))
```
They shouldn't have to be *both* secretive and provoked on scan to give a warning, should they?
Or, even, secretive ships just shouldn't send this message at all? If you're trying to keep something a secret, you may not necessarily wish to advertise it.

File: source/Ship.cpp [hunk 2]
```diff
@@ -1435,6 +1436,37 @@ string Ship::GetHail(map<string, string> &&subs) const
...
+		return false;
+
+	// Player ships shouldn't send hails.
+	const Government *gov = GetGovernment();
+	if(!gov || IsYours())
+		return false;
+
+	// Make sure this ship is able to send a hail.
+	if(IsDisabled() || !Crew()
+			|| Cloaking() >= 1. || GetPersonality().IsMute())
```
Comment (warp-core) [inline source/Ship.cpp:1456]:
```suggestion
	if(IsDisabled() || !Crew() || Cloaking() >= 1. || GetPersonality().IsMute())
```
By my count, this line is exactly 80 characters so doesn't need to be split.

~~Also, `automated hail`s sent when a ship has no crew when?~~

File: source/Ship.cpp (hunk 1 above)
Comment (Hurleveur) [reply source/Ship.cpp:2707]: I think it works best with both tbh
Not all provoked on ships would say smth (also not the idea of this PR to change this behavior)

That said, if thats the only blocker of the PR do feel free to change it

Comment (Hurleveur) [reply source/Ship.cpp:2706]: Smugglers for instance would be secretive but would not say anything because not provoked on scan

Comment (Hurleveur) [reply source/Ship.cpp:2706]:
```suggestion
		if(target->GetGovernment()->IsProvokedOnScan() && target->CanSendHail(player))
```
This is acceptable

Comment (warp-core) [reply source/Ship.cpp:2706]: Requiring both means that something that is provoked on scan but isn't secretive wouldn't give the warning hails.

Comment (Hurleveur) [reply source/Ship.cpp:2706]: Alr then, only use case for provoked on scan is the remnants anyway and it works for them
Lets just do this and change it later if needed

**Review by warp-core — APPROVED** [review]

**Pull request merged**